File: 03-hard/0224-prisoner-or-citizen/0224-prisoner-or-citizen.cpp
/*
  Copyright (C) 2015 Daniel Rodriguez

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

/*
  This is a "Point in a Polygon" problem with boundaries counting as part of
  the polygon

  References
    - http://www.ics.uci.edu/~eppstein/161/960307.html
    - http://geomalgorithms.com/a03-_inclusion.html
    - http://geomalgorithms.com/a01-_area.html

  To be investigated for improvement:
    - http://www.inf.usi.ch/hormann/papers/Hormann.2001.TPI.pdf
    - https://www.engr.colostate.edu/~dga/dga/papers/point_in_polygon.pdf

  Although Raycasting is also a potential solution, for a problem with the
  given constraints the Winding Number Algorithm should be better because it
  provides accuracy by using only integer operations.
*/

// Headers for test case
#include <fstream>
#include <iostream>
#include <string>
#include <sstream>

// Headers for the implementation
#include <algorithm>
#include <iterator>
#include <vector>
#include <forward_list>


class Point_t {
public:
    size_t x;
    size_t y;

    Point_t(const size_t &x, const size_t &y) : x{x}, y{y} {}
};

using Vertex_t = Point_t;


template <typename Seq>
auto
is_in_poly(const Seq &vs, const Point_t &p)
//    -> bool
{
    /*
      Input:
        - vs: A sequence of vertex containing the definition of the polygon
        - p: The point to be tested
      Output:
        - bool: true -> It's inside polygon, false -> not in the polygon

      Because being on the "edges" of the polygon is considered being in the
      polygon, the algorithm exits immediately if such condition is detected at
      any stage
    */

    ssize_t wn = 0;    // the  winding number counter

    // loop through all edges of the polygon
    for (auto vs0=vs.begin(), vs1=std::next(vs0); vs1 != vs.end(); vs0++, vs1++) {
        auto v0 = *vs0;
        auto v1 = *vs1;

        auto OnLineX = [&] () -> bool {
            // p is within the x coordinates of v0/v1
            if(v0.x <= v1.x)
                return v0.x <= p.x and p.x <= v1.x;
            return v0.x >= p.x and p.x >= v1.x;
        };

        auto LeftOrRight = [&] () -> ssize_t {
            // Must set return type or else "unsigned" will be auto chosen
            // Left/Right -  < 0 Right, 0 = On Line, > 0 Left
            return ((v1.x - v0.x) * (p.y - v0.y)) - ((v1.y - v0.y) * (p.x - v0.x));
        };

        if(v1.y >= p.y && p.y > v0.y) {   // upwards crossing
            auto lor = LeftOrRight();
            if(lor > 0)
                wn++;
            else if(lor == 0)
                return true;
        } else if(v0.y >= p.y && p.y > v1.y) {   // downwards crossing
            auto lor = LeftOrRight();
            if(lor < 0)  // Right of line
                wn--;
            else if(lor == 0)
                return true;
        } else if(v0.y == p.y && OnLineX())
            return true;
    }
    return wn != 0;
}


struct SeparatorReader: std::ctype<char>
{
    SeparatorReader(const std::string &seps):
        std::ctype<char>(get_table(seps), true) {}

    std::ctype_base::mask const *get_table(const std::string &seps) {
        auto rc = new std::ctype_base::mask[std::ctype<char>::table_size]();
        for(auto &sep: seps)
            rc[sep] = std::ctype_base::space;
        return &rc[0];
    }
};


int
main(int argc, char *argv[]) {
    std::ifstream stream(argv[1]);
    std::string line;

    std::stringstream sscoord;
    sscoord.imbue(std::locale(sscoord.getloc(), new SeparatorReader(" ,|\n")));

    while (std::getline(stream, line)) {
        sscoord.str(line); sscoord.clear();  // set str & clear any eol/eof

        // Parse all coordinates (vertex and point)
        std::vector<Point_t> coords;
        for (int x=0, y=0; sscoord >> x and sscoord >> y;)
            coords.push_back(Point_t(x, y));

        // extract point and replace with initial vertex to close polygon
        auto pt = coords.back();
        coords[coords.size() - 1] = coords.front();

        std::cout << (is_in_poly(coords, pt) ? "Prisoner" : "Citizen") << std::endl;
    }
    return 0;
}

File: 03-hard/0224-prisoner-or-citizen/0224-prisoner-or-citizen.py
#!/usr/bin/env python
# -*- coding: utf-8; py-indent-offset:4 -*-
###############################################################################
#
# Copyright (C) 2015 Daniel Rodriguez
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
###############################################################################
from __future__ import (absolute_import, division, print_function,)
#                        unicode_literals)

import collections
import sys

# Small Py2/3 compatibility layer
if sys.version_info.major == 2:
    MAXINT = sys.maxint
    MININT = -sys.maxint - 1

    import itertools
    filter = itertools.ifilter
    map = itertools.imap
    range = xrange
    zip = itertools.izip

else:  # >= 3
    MAXINT = sys.maxsize
    MININT = -sys.maxsize - 1


class Point(object):
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __str__(self):
        return '%d %d' % (self.x, self.y)


def is_in_poly(vs, p):
    # Returns True if "on line" or winding number is not 0 (in polygon)
    wn = 0  # winding number

    for i in range(len(vs) - 1):
        v0, v1 = vs[i], vs[i + 1]

        # the 2 following functions operate with closures for direct access to
        # v0, v1 and p
        #
        # Return if point is on-line
        def on_line_x():
            if v0.x <= v1.x:
                return v0.x <= p.x and p.x <= v1.x

            return v0.x >= p.x and p.x >= v1.x

        # return if point is left or right of v0->v1
        # The result is interpreted with regards to v0->v1 vertical direction
        def left_or_right():
            return ((v1.x - v0.x) * (p.y - v0.y)) - ((v1.y - v0.y) * (p.x - v0.x))

        if v1.y >= p.y and p.y > v0.y:  # v0 -> v1 upwards
            lor = left_or_right()
            if lor > 0:
                wn += 1
            elif lor == 0:
                return True
        elif v0.y >= p.y and p.y > v1.y:  # v1 -> v0 downwards
            lor = left_or_right()
            if lor < 0:
                wn -= 1
            elif lor == 0:
                return True
        elif v0.y == p.y and on_line_x():  # v0/v1 on same vertical as point
            return True

    return wn != 0


if __name__ == '__main__':
    test_cases = open(sys.argv[1], 'r')
    for line in test_cases:
        ls = line.rstrip('\r\n')

        # Parse the coordinates - separate vertex from point
        vxys, _, pxy = ls.rpartition('|')
        # Parse the vertex
        vcoords = list()
        for vxy in vxys.split(','):
            x, y = vxy.split()
            vcoords.append(Point(int(x), int(y)))

        # Append 1st as last to close the polygon
        vcoords.append(vcoords[0])  # close the polygon

        # Parse the point
        x, y = pxy.split()
        pt = Point(int(x), int(y))

        print("Prisoner" if is_in_poly(vcoords, pt) else "Citizen")

    test_cases.close()
